// @carto/airship-components/dist/cjs/as-y-axis.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const core = require('./core-894cddb9.js');
require('./readable-number-3954c8a2.js');
require('./index-4a52f739.js');
require('./index-f86f0d00.js');
const yAxis_service = require('./y-axis.service-3c96e9a0.js');

const YAxis = class {
    constructor(hostRef) {
        core.registerInstance(this, hostRef);
        /**
         * Lower limit of the axis
         *
         * @type {number}
         * @memberof YAxis
         */
        this.from = 0;
        /**
         * Upper limit of the axis
         *
         * @type {Number[]}
         * @memberof YAxis
         */
        this.to = 0;
        /**
         * Use this attribute to decide if the widget should be rerendered on window resize
         * Defaults to true
         */
        this.responsive = true;
        this._resizeListener = this._resizeListener.bind(this);
    }
    componentWillLoad() {
        addEventListener('resize', this._resizeListener);
    }
    componentDidUnload() {
        removeEventListener('resize', this._resizeListener);
    }
    render() {
        const element = this.element.previousElementSibling;
        const scale = [this.from, this.to];
        yAxis_service.yAxisService.renderYAxis(element, scale);
    }
    _resizeListener() {
        if (this.responsive) {
            this.element.forceUpdate();
        }
    }
    get element() { return core.getElement(this); }
    static get style() { return ".y-axis{--widget-axis-text-color:var(--as--color--type-01,#2c2c2c);--widget-axis-line-color:var(--as--color--ui-05,#b3b3b3)}.y-axis .tick text{width:30px;fill:var(--widget-axis-text-color);white-space:pre}.y-axis .tick line{stroke:var(--widget-axis-line-color);opacity:.1}.y-axis .tick line.zero{opacity:.5}.y-axis .domain{display:none}"; }
};

exports.as_y_axis = YAxis;
